elegant/edge_detection.py

```diff
@@ -0,0 +1,203 @@
+    to the edges of the worm)
+
+    Parameters:
+        image: ndarray of the straightened worm image (typically either top or bottom half)
+        avg_width_tck: width spline defining the average distance from the centerline
+            to the worm edges (This is taken from the pca things we did earlier)
+        ggm_sigma, sig_per, sig_growth_rate, alpha, mcp_alpha: hyperparameters for 
+            the edge-detection scheme
+    
+    Returns:
+        route: tuple of x,y positions of the identfied edges
+    """
+
+    #down sample the image
+    image_down = pyramid.pyr_down(image, downscale=2)
+
+    #get the gradient
+    gradient = ndimage.filters.gaussian_gradient_magnitude(image_down, 1)
+    top_ten = np.percentile(gradient,  61)
+    gradient = sigmoid(gradient, gradient.min(), top_ten, gradient.max(), 2)
+    gradient = gradient.max()-abs(gradient)
+
+    #penalize finding edges near the centerline or outside of the avg_width_tck
+    #since the typical worm is fatter than the centerline and not huge
+    #Need to divide by 2 because of the downsampling
+    pen_widths = (interpolate.spline_interpolate(avg_width_tck, image_down.shape[0]))
+    #pen_widths = pen_widths/2
+    distance_matrix = abs(np.subtract.outer(pen_widths, np.arange(0, image_down.shape[1])))
+    #distance_matrix = np.flip(abs(np.subtract.outer(pen_widths, np.arange(0, image_down.shape[1]))), 1)
+    penalty = alpha*(distance_matrix)
+    new_costs = gradient+penalty
+    
+    #set start and end points for the traceback
+    start = (0, int(pen_widths[0]))
+    end = (len(pen_widths)-1, int(pen_widths[-1]))
+    #start = (0, int((image_down.shape[1]-1)-pen_widths[0]))
+    #end = (len(pen_widths)-1, int((image_down.shape[1]-1)-pen_widths[-1]))
+    #start = (0,0)
+    #end = (len(pen_widths)-1, 0)
+
+    #begin edge detection
+    offsets= [(1,-1),(1,0),(1,1)]
+    mcp = Smooth_MCP(new_costs, mcp_alpha, offsets=offsets)
+    mcp.find_costs([start], [end])
+    route = mcp.traceback(end)
+
+    x,y = np.transpose(route)
+    #print("x: ", len(x), "y:", len(y))
+    #print(route)
+    #y = image_down.shape[1] - y
+    return (x*2, y*2) #multiply by 2 to account for downsampling
+
+def sigmoid(gradient, min, mid, max, growth_rate):
+    '''Apply the sigmoid function to the gradient.
+
+    Parameters:
+        gradient: array of the gradient of the image
+        min: lower asymptote of the sigmoid function
+        mid: midpoint of the sigmoid function (ie. the point that is halfway between
+            the lower and upper asymptotes)
+        max: upper asymptote
+        growth_rate: growth rate of the sigmoid function
+    '''
+    return min+((max-min)/(1+np.exp(-(growth_rate)*(gradient-mid))))
+
+def scale_image(image):
+    mode = np.bincount(image.flat)[1:].argmax()+1
+    bf = image.astype(np.float32)
+    bf -= 200
+    bf *= (24000-200) / (mode-200)
+    bf8 = colorize.scale(bf, min=600, max=26000, gamma=0.72, output_max=255).astype(np.float32)
```

Comment:
A little weird to scale it from 0 to 255, since you are using the image as floating-point. Also, the output of `colorize.scale` should already be floating-point, no? Or are you taking it from `float64` to `float32` here?